File: ecosystems/luarocks/aitextcleaner_formatting_residue.lua
local block_tags = {
  address = true, article = true, blockquote = true, br = true, div = true,
  li = true, p = true, pre = true, section = true,
}

local function decode_entities(value)
  local entities = {amp = "&", apos = "'", gt = ">", lt = "<", nbsp = " ", quot = '"'}
  return (value:gsub("&([%a]+);", function(entity)
    return entities[entity:lower()] or "&" .. entity .. ";"
  end):gsub("&#39;", "'"):gsub("&#x27;", "'"))
end

local function clean(text)
  assert(type(text) == "string", "text must be a string")
  local cleaned = text:gsub("<[^>]*>", function(tag)
    local name = tag:match("^<%s*/?%s*([%a][%w-]*)")
    if name and block_tags[name:lower()] then return "\n" end
    return ""
  end)
  cleaned = decode_entities(cleaned)
  cleaned = cleaned:gsub("“", '"'):gsub("”", '"'):gsub("‘", "'"):gsub("’", "'")
    :gsub("–", "-"):gsub("—", "-"):gsub("…", "...")
    :gsub("[ \t]+\n", "\n"):gsub("\n\n\n+", "\n\n")
  return cleaned:match("^%s*(.-)%s*$") or ""
end

return {clean = clean}

File: ecosystems/pub/lib/aitextcleaner_formatting_residue.dart
const _blockTag = r'<\/?(?:address|article|blockquote|br|div|li|p|pre|section)\b[^>]*>';
final _blockTagPattern = RegExp(_blockTag, caseSensitive: false);
final _htmlTagPattern = RegExp(r'<[^>]*>');

String _decodeEntities(String value) {
  return value.replaceAllMapped(RegExp(r'&(#x?[0-9a-f]+|amp|apos|gt|lt|nbsp|quot);', caseSensitive: false), (match) {
    final entity = match.group(1)!.toLowerCase();
    if (entity == 'amp') return '&';
    if (entity == 'apos' || entity == '#39' || entity == '#x27') return "'";
    if (entity == 'gt') return '>';
    if (entity == 'lt') return '<';
    if (entity == 'nbsp') return ' ';
    if (entity == 'quot') return '"';
    final radix = entity.startsWith('#x') ? 16 : 10;
    final digits = entity.substring(entity.startsWith('#x') ? 2 : 1);
    return String.fromCharCode(int.parse(digits, radix: radix));
  });
}

String cleanFormattingResidue(String text) {
  var cleaned = text.replaceAll(_blockTagPattern, '\n').replaceAll(_htmlTagPattern, '');
  cleaned = _decodeEntities(cleaned)
      .replaceAll('“', '"')
      .replaceAll('”', '"')
      .replaceAll('‘', "'")
      .replaceAll('’', "'")
      .replaceAll('–', '-')
      .replaceAll('—', '-')
      .replaceAll('…', '...')
      .replaceAll(RegExp(r'[ \t]+\n'), '\n')
      .replaceAll(RegExp(r'\n{3,}'), '\n\n');
  return cleaned.trim();
}
